api/src/main/java/io/minio/ComposeSource.java

```diff
@@ -0,0 +1,188 @@
+    this(bucketName, objectName, null, null, null, null, null);
+  }
+
+  /**
+   * Create new ComposeSource for given bucket, object, offset and length.
+   */
+  public ComposeSource(String bucketName, String objectName, Long offset, Long length)
+    throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, null, null, null);
+  }
+
+  /**
+   * Create new ComposeSource for given bucket, object, offset, length and headerMap.
+   */
+  public ComposeSource(String bucketName, String objectName, Long offset, Long length,
+      Map<String, String> headerMap) throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, headerMap, null, null);
+  }
+
+  /**
+   * Create new ComposeSource for given bucket, object, offset, length, headerMap and CopyConditions.
+   */
+  public ComposeSource(String bucketName, String objectName, Long offset, Long length,
+      Map<String, String> headerMap, CopyConditions copyConditions) throws InvalidArgumentException {
+    this(bucketName, objectName, offset, length, headerMap, copyConditions, null);
+  }
+
+  /**
+   * Creates new ComposeSource for given bucket, object, offset, length, headerMap, CopyConditions
+   * and server side encryption.
+   *
+   * @throws InvalidArgumentException    upon invalid value is passed to a method.
+   */
+  public ComposeSource(String bucketName, String objectName, Long offset, Long length, Map<String, String> headerMap,
+      CopyConditions copyConditions, ServerSideEncryption sse) throws InvalidArgumentException {
+    if (bucketName == null) {
+      throw new InvalidArgumentException("Source bucket name cannot be empty");
+    }
+
+    if (objectName == null) {
+      throw new InvalidArgumentException("Source object name cannot be empty");
+    }
+
+    if (offset != null && offset < 0) {
+      throw new InvalidArgumentException("Offset cannot be negative");
+    }
+
+    if (length != null && length < 0) {
+      throw new InvalidArgumentException("Length cannot be negative");
+    }
+
+    if (length != null && offset == null) {
+      offset = 0L;
+    }
+
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    this.offset = offset;
+    this.length = length;
+    this.headerMap = headerMap;
+    this.copyConditions = copyConditions;
+    this.sse = sse;
+  }
+
+  /**
+   * Constructs header  .
+   *
+   */
+  public void buildHeaders(long objectSize, String etag) throws InvalidArgumentException {
+    if (offset != null && offset >= objectSize) {
+      throw new InvalidArgumentException("source " + bucketName + "/" + objectName + ": offset " + offset
+        + " is beyond object size " + objectSize);
+    }
+
+    if (length != null) {
+      if (length >= objectSize) {
+        throw new InvalidArgumentException("source " + bucketName + "/" + objectName + ": length " + length
+          + " is beyond object size " + objectSize);
+      }
+
+      if (offset + length > objectSize) {
```

Comment:
Can you explain why we need to error out if `offset+length == objectSize`?

Reply:
Generally `offset` is an integer (including negative numbers) refers an index and `length` is a non-negative integer. If `size` is 10, `offset = 4` and `length = 2` refers two bytes from 4th index whereas `offset = -4` and `length = 2` refers two bytes from end of 4th index i.e. `offset = 6`. `length=0` is a legitimate value which refers zero bytes.

In `ComposeSource`
1. If `offset`/`length` is non-`null`, it must be >= 0 i.e. we do not support negative offset.

2. `offset`/`length` can be `null`.  `offset=null` means `offset=0` and `length=null` means `length=size-offset`. This validation is done in the constructor.

3. There is no way to get object size (`size`) in the constructor to calculate `length=size-offset` when `length=null` because the size is unknown.

4. At the time of `buildHeaders()`, `size` and `etag` are passed from `composeObject()` in `MinioClient` because `MinioClient` class is S3 client fetches `ObjectStat` for given `ComposeSource`.

5. `buildHeaders()` should calculate `length=size-offset` when `length=null`.  @sinhaashish This should be fixed. Note that its not required to set it to `this.length`.

6. S3 accepts byte range not offset/length and byte range are calculated like below
  * When `offset=null` and `length=null`, no byte range header is set.
  * When `offset=null` and `length=non-null`, byte range is `0-length`.
  * When `offset=non-null` and `length=null`, byte range is `offset-`.
  * When `offset=non-null` and `length=non-null`, byte range is `offset-length`.

Reply:
Adding a test case for this  using the object at `play/compose-object/15Lakh.csv`.
![Test](https://user-images.githubusercontent.com/13465383/64530778-93e7b080-d32b-11e9-8e07-283ddccdd09b.png)